### PR TITLE
Optimize checks for shared files

### PR DESCRIPTION
**Pull request by mathiascode**
We can avoid looping through the list of files in a folder.

### CODE REVIEW

**Comment by mathiascode** [comment]
This could probably be merged with `virtual2real`. Need to do more testing and cleanups.

**Comment by mathiascode** [comment]
Should be ready for testing.

**Comment by slook** [comment]
There is no difference in behaviour nor speed when re-scanning shares that are not changed:

_master ~7 seconds:_
```
2022-08-02 23:15:00 Rescanning shares…
2022-08-02 23:15:00 2895 folders found before rescan, rebuilding…
2022-08-02 23:15:01 Error while scanning folder /media/TIX/SoulSeek/Complete/[1000+02] Pyrrhon / Untitled: [Errno 2] No such file or directory: b'/media/TIX/SoulSeek/Complete/[1000+02] Pyrrhon / Untitled'
2022-08-02 23:15:01 Error while scanning folder /media/TIX/SoulSeek/Complete/Deltron 3030 / Event 2 (2013): [Errno 2] No such file or directory: b'/media/TIX/SoulSeek/Complete/Deltron 3030 / Event 2 (2013)'
2022-08-02 23:15:07 Rescan complete: 2895 folders found
```

_optimize-file-shared ~7 seconds:_
```
2022-08-02 23:16:00 Rescanning shares…
2022-08-02 23:16:00 2895 folders found before rescan, rebuilding…
2022-08-02 23:16:01 Error while scanning folder /media/TIX/SoulSeek/Complete/[1000+02] Pyrrhon / Untitled: [Errno 2] No such file or directory: b'/media/TIX/SoulSeek/Complete/[1000+02] Pyrrhon / Untitled'
2022-08-02 23:16:01 Error while scanning folder /media/TIX/SoulSeek/Complete/Deltron 3030 / Event 2 (2013): [Errno 2] No such file or directory: b'/media/TIX/SoulSeek/Complete/Deltron 3030 / Event 2 (2013)'
2022-08-02 23:16:07 Rescan complete: 2895 folders found
```

**Comment by slook** [comment]
An upload was successful, though I doubt the performance of those lookups would be possible to measure unless you add some debug logging with timestamps.

File: pynicotine/shares.py [hunk 1]
```diff
@@ -485,55 +507,61 @@ def init_shares(self):
...
-                return real
+        return None
 
-            if path.startswith(virtual + os.sep):
-                realpath = real + path[len(virtual):]
-                return realpath
+    def virtual2real(self, virtual_path, user=None):
 
-        return "__INTERNAL_ERROR__" + path
+        if user is None:
```
Comment (slook) [inline pynicotine/shares.py:546]: Brief # comment on the purpose of this condition being here.

File: pynicotine/transfers.py [hunk 2]
```diff
@@ -1180,9 +1178,9 @@ def file_upload_init(self, msg):
             if upload in self.transfer_request_times:
                 del self.transfer_request_times[upload]
 
-            real_path = self.core.shares.virtual2real(filename)
+            real_path = self.core.shares.virtual2real(filename, username)
```
Comment (slook) [inline pynicotine/transfers.py:1181]: Why is it necessary for the `username` argument to be parsed from here, but is not required from userbrowse?

File: pynicotine/shares.py [hunk 3]
```diff
@@ -485,55 +507,61 @@ def init_shares(self):
...
+            return real_path
 
-            if path == virtual:
-                return real
+        return None
 
-            if path.startswith(virtual + os.sep):
-                realpath = real + path[len(virtual):]
-                return realpath
+    def virtual2real(self, virtual_path, user=None):
```
Comment (slook) [inline pynicotine/shares.py:544]: This method now contains code that seems unrelated to the name of this function.

**Comment by mathiascode** [comment]
Rescanning shares shouldn't be any different. The main area this PR touches is upload queue requests, which would be ideal to test by queuing a lot of files at once with the queue limit disabled.

File: pynicotine/shares.py [hunk 4]
```diff
@@ -485,55 +507,61 @@ def init_shares(self):
...
+                continue
 
-            if path.startswith(real):
-                virtualpath = virtual + '\\' + path[len(real):]
-                return virtualpath
+            real_path = real_folder_path + virtual_path[len(virtual_name):]
 
-        return "__INTERNAL_ERROR__" + path
+            if real_path != os.path.abspath(real_path):
+                # Protect against directory traversal
```
Comment (slook) [inline pynicotine/shares.py:532]: What circumstance can lead to this condition?

**Comment by slook** [comment]
> The main area this PR touches is upload queue requests

Thanks for the clarification, the purpose of the functions in this complex module are difficult for me to understand.

**Comment by slook** [comment]
Is this PR still relevant or does it need to be closed now?

**Comment by mathiascode** [comment]
Closing for now, I don't feel confident in merging this right now.
